doc/design/build_system/go.md

```diff
@@ -0,0 +1,98 @@
+# Design Doc: `go_{library,binary,test}`
+
+## Concerns
+
+1. Need to build Go libraries callable from Go and from C.
+
+   For usual Go libraries, the bulding command line is as
+
+   ```bash
+   go build foo.go bar.go -o libfoobar.a
+   ```
+
+   For Go libraries callable from C/C++, the command line is
+
+   ```bash
+   go build -buildmode=c-archive foo.go bar.go -o libstatic.a
+   ```
+
+   or for shared libraries:
+
+   ```bash
+   go build -buildmode=c-shared foo.go bar.go -o libdynamic.so
+   ```
+
+   and `foo.go`, `bar.go`, etc must start with a line `package main`,
+   which defines all symbols in special pacakge `main`.  There also
+   must be a `func main`, which could have empty body.
+
+1. Need to support building-in-Docker.
+
+   We are going to support two ways to building -- (1) in Ubuntu, and
+   (2) in Docker container whose base image is Ubuntu.
+
+   The challenge is (2), because to build in Docker, we run the
+   development image as:
+
+   ```bash
+   git clone https://github.com/PaddlePaddle/Paddle -o paddle
+   cd paddle
+   docker run -v $PWD:/paddle paddle:dev
+   ```
+
+   which maps the local repo to `/paddle` in the container.
+
+   This assumes that all Go code, including third-party packages, must
+   be in the local repo.  Actually, it assumes that `$GOPATH` must be
+   in the local repo.  This would affect how we write `import`
+   statemetns, and how we maintain third-party packages.
+
+
+## A Solution
+
+This might not be the optimal solution.  Comments are welcome.
+
+### Directory structure
+
+We propose the following directory structure:
+
+```
+https://github.com/PaddlePaddle/Paddle
+                                  ↓ git clone
+                         ~/work/paddle/go/pkg1/foo.go
+                                         /pkg2/bar.go
+                                         /cmd/cmd1/wee.go
+                                         /cmd/cmd2/qux.go
+                                         /github.com/someone/a_3rd_party_pkg (Git submodule to a 3rd-party pkg)
+                                         /golang.org/another_3rd_party_pkg (Git submodule to another one)
+                                      /build/go ($GOPATH, required by Go toolchain)
+                                               /src (symlink to ~/work/paddle/go/)
+                                               /pkg (libraries built by Go toolchain)
+                                               /bin (binaries bulit by Go toolchain)
+```
+
+Above figure explains how we organize Paddle's Go code:
+
+1. Go source code in Paddle is in `/go` of the repo.
+1. Each library package is a sub-directory under `/go`.
+1. Each (executable) binary package is a sub-directory under
+   `/go/cmd`.  This is the source tree convention of Go itself.
+1. Each 3rd-party Go package is a Git submodule under `/go`.
```

Comment:
There are many pros for using this submodule based solution:
1. dependency version management is easy (by git hash).
1. can modify dependency when doing experiments in the local filesystem.

Here are the cons that I can think of:
1. Go import path support version control system like "git", "hg", "svn", (`go help importpath`), but using submodule is limited on git.
1. need to manually add submodule for every dependency, when the dependency is already specified in the go source code.

I vote for the approach of set GOPATH to ./build/go and use `go get` to fetch dependency.
1. more go idiomatic.
1. can still modify dependency when doing experiments in the local filesystem.
1. if we want dependency management, we can use go community's spec and solutions: https://github.com/golang/go/wiki/PackageManagementTools